kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/PulsarStorageProducerIdManagerImpl.java

```diff
@@ -0,0 +1,210 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.streamnative.pulsar.handlers.kop.coordinator.transaction;
+
+import java.math.BigInteger;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Collectors;
+import java.util.stream.LongStream;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.pulsar.client.api.Message;
+import org.apache.pulsar.client.api.MessageId;
+import org.apache.pulsar.client.api.Producer;
+import org.apache.pulsar.client.api.ProducerAccessMode;
+import org.apache.pulsar.client.api.PulsarClient;
+import org.apache.pulsar.client.api.Reader;
+
+/**
+ * ProducerIdManager is the part of the transaction coordinator that provides ProducerIds in a unique way
+ * such that the same producerId will not be assigned twice across multiple transaction coordinators.
+ *
+ * ProducerIds are managed via a Pulsar non-partitioned topic.
+ */
+@Slf4j
+public class PulsarStorageProducerIdManagerImpl implements ProducerIdManager {
+    static final int BLOCK_SIZE = 1000;
+    private final AtomicLong nextId = new AtomicLong(0);
+    private final String topic;
+    private final int blockSize;
+    private final PulsarClient pulsarClient;
+    private CompletableFuture<Reader<byte[]>> reader;
+    private CompletableFuture<Void> currentReadHandle;
+    private final ConcurrentLinkedQueue<Long> availableIdsLocally = new ConcurrentLinkedQueue<>();
+
+    private synchronized CompletableFuture<Reader<byte[]>> ensureReaderHandle() {
+        if (reader == null) {
+            reader = pulsarClient.newReader()
+                    .topic(topic)
+                    .startMessageId(MessageId.earliest)
+                    .startMessageIdInclusive()
+                    .readCompacted(true)
+                    .createAsync();
+        }
+        return reader;
+    }
+
+    private CompletableFuture<Void> readNextMessageIfAvailable(Reader<byte[]> reader) {
+        return reader
+                .hasMessageAvailableAsync()
+                .thenCompose(hasMessageAvailable -> {
+                    if (hasMessageAvailable == null
+                            || !hasMessageAvailable) {
+                        return CompletableFuture.completedFuture(null);
+                    } else {
+                        CompletableFuture<Message<byte[]>> opMessage = reader.readNextAsync();
+                        return opMessage.thenCompose(msg -> {
+                            byte[] value = msg.getValue();
+                            long newId = new BigInteger(value).longValue();
+                            log.debug("{} Read {} from {}", this, newId, msg.getMessageId());
```

Comment:
Please add the `if (log.isDebugEnabled())` block.

Reply:
sure